kimchi/src/tests/foreign_field_add.rs

```diff
@@ -0,0 +1,1284 @@
+    0xAA, 0xAA, 0xAA, 0xAA, 0xA9, 0xBA, 0xAA, 0xAA, 0xAA, 0xAA, 0xAA, 0x00, 0x00, 0x00, 0x00, 0x00,
+];
+
+// BigEndian -> 00000000 00000000 00000000 00000000 FFFFFFFF FFFFFFFF FFFFFFFE FFFFFC2F
+/// Bottom half of the foreign modulus
+static FOR_MOD_BOT: &[u8] = &[
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFE, 0xFF, 0xFF, 0xFC, 0x2F,
+];
+
+// BigEndian -> FFFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF 00000000 00000000 00000000 00000000
+/// Top half of the foreign modulus
+static FOR_MOD_TOP: &[u8] = &[
+    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+];
+
+/// Value that performs a + - 1 low carry when added to [MAX]
+static NULL_CARRY_LO: &[u8] = &[0x01, 0x00, 0x00, 0x03, 0xD2];
+
+/// Value that performs a + - 1 middle carry when added to [MAX]
+static NULL_CARRY_MI: &[u8] = &[
+    0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
+];
+
+/// Value that performs two + - 1 carries when added to [MAX]
+static NULL_CARRY_BOTH: &[u8] = &[
+    0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x03, 0xD2,
+];
+/// The zero byte
+static ZERO: &[u8] = &[0x00];
+
+/// The one byte
+static ONE: &[u8] = &[0x01];
+
+fn create_test_constraint_system_ffadd(
+    num: usize,
+    modulus: BigUint,
+) -> ConstraintSystem<PallasField> {
+    let (mut next_row, mut gates) = CircuitGate::<PallasField>::create_foreign_field_add(0, num);
+
+    // Temporary workaround for lookup-table/domain-size issue
+    for _ in 0..(1 << 13) {
+        gates.push(CircuitGate::zero(Wire::new(next_row)));
+        next_row += 1;
+    }
+
+    ConstraintSystem::create(gates)
+        .foreign_field_modulus(&Some(modulus))
+        .build()
+        .unwrap()
+}
+
+#[test]
+// Add zero to zero. This checks that small amounts also get packed into limbs
+fn test_zero_add() {
+    let foreign_modulus = BigUint::from_bytes_be(FOREIGN_MOD);
+    let cs = create_test_constraint_system_ffadd(1, foreign_modulus.clone());
+
+    let left_input = BigUint::from_bytes_be(ZERO);
+    let right_input = BigUint::from_bytes_be(ZERO);
+
+    let witness = create_witness(vec![left_input, right_input], vec![ADD], foreign_modulus);
+
+    for row in 0..=18 {
+        assert_eq!(
+            cs.gates[row].verify_witness::<Vesta>(
+                row,
+                &witness,
+                &cs,
+                &witness[0][0..cs.public].to_vec()
+            ),
+            Ok(())
+        );
+    }
+
+    for row in 16..=18 {
+        assert_eq!(
+            cs.gates[row].verify::<Vesta>(row, &witness, &cs, &[]),
+            Ok(())
+        );
+    }
+}
+
+#[test]
+// Adding terms that are zero modulo the foreign field
+fn test_zero_sum_foreign() {
```

Comment:
A lot of these test seem to have the same basic structure.. I wonder if there's a way you can simplify them in order to remove a lot of lines of code?  For example,

`check(FOREIGN_MOD, FOR_MOD_BOT, FOR_MOD_TOP, { result check })`